File: crm2/klemco_cs/events/item.py
# Item server-side events — BRD v1.3
#   CR-18 / BR-KM-02  New KM-master items require TRIPLE approval before they can be enabled:
#                     CS Supervisor + KM Plant Head + Supply Chain Lead.

import frappe
from frappe import _

# approval field -> role allowed to grant it
KM_APPROVALS = (
    ("custom_km_approved_cs_supervisor", "CS Supervisor"),
    ("custom_km_approved_plant_head", "KM Plant Head"),
    ("custom_km_approved_supply_chain", "Supply Chain Lead"),
)


def validate(doc, method=None):
    if not doc.get("custom_km_managed"):
        doc.custom_km_approval_status = None
        return

    _enforce_role_for_each_approval(doc)

    all_approved = all(doc.get(field) for field, _role in KM_APPROVALS)
    doc.custom_km_approval_status = "Approved" if all_approved else "Pending Approvals"

    # A KM-managed item may not be Active (disabled = 0) until all three approve.
    if not all_approved and not doc.get("disabled"):
        frappe.throw(_(
            "KM-managed Item requires triple approval — CS Supervisor + KM Plant Head + "
            "Supply Chain Lead — before it can be enabled (BR-KM-02 / CR-18). Keep it "
            "'Disabled' until all three approvals are granted."
        ))


def _enforce_role_for_each_approval(doc):
    before = doc.get_doc_before_save()
    roles = set(frappe.get_roles())
    privileged = "System Manager" in roles

    for field, role in KM_APPROVALS:
        was = before.get(field) if before else 0
        now = doc.get(field)
        if now and not was and not privileged and role not in roles:
            frappe.throw(_(
                "Only the {0} role can grant the '{0}' approval for KM-managed items (BR-KM-02)."
            ).format(role))

File: crm2/klemco_cs/public/js/item.js
// Item — Klemco CS client script (BRD v1.3)
//   CR-18 / BR-KM-02  KM-managed items: show the triple-approval progress and gate each check
//                     so only the matching role can grant its approval.

const KM_APPROVALS = [
    { field: 'custom_km_approved_cs_supervisor', role: 'CS Supervisor', label: 'CS Supervisor' },
    { field: 'custom_km_approved_plant_head', role: 'KM Plant Head', label: 'KM Plant Head' },
    { field: 'custom_km_approved_supply_chain', role: 'Supply Chain Lead', label: 'Supply Chain Lead' },
];

frappe.ui.form.on('Item', {
    refresh(frm) {
        if (!frm.doc.custom_km_managed) return;

        const roles = frappe.user_roles || [];
        // Only let a user tick the approval that matches their role.
        KM_APPROVALS.forEach(a => {
            const allowed = roles.includes(a.role) || roles.includes('System Manager');
            frm.set_df_property(a.field, 'read_only', allowed ? 0 : 1);
        });

        const done = KM_APPROVALS.filter(a => frm.doc[a.field]).map(a => a.label);
        const pending = KM_APPROVALS.filter(a => !frm.doc[a.field]).map(a => a.label);
        if (pending.length) {
            frm.dashboard.set_headline_alert(
                __('KM item triple approval pending: {0}. Approved: {1}.',
                   [pending.join(', '), done.join(', ') || '—']),
                'orange'
            );
        } else {
            frm.dashboard.set_headline_alert(__('KM item fully approved (CS Supervisor + KM Plant Head + Supply Chain Lead).'), 'green');
        }
    },

    custom_km_managed(frm) {
        frm.refresh();
    },
});
